Add JS usage examples for building and using contract specifications (#662)
stellar/stellar-docs: docs/smart-contracts/guides/conventions/work-contractspec-js.mdx

draft: true
---

### Guide: Work with ContractSpecs in JavaScript

#### Introduction

Soroban smart contracts, integral to the Stellar network, utilize detailed contract specifications to define their behavior and interactions. A contract specification outlines the functions, parameters, and events within a smart contract, guiding developers in how to interact with these contracts through JavaScript. This guide will walk you through the process of creating a contract client and invoking smart contracts using JavaScript.

#### Prerequisites

To follow this guide, you will need:

- Node.js (version 18 or later)
- `@stellar/stellar-sdk` package
- A working knowledge of JavaScript and basic understanding of smart contracts

### Information on the Contract Spec

> The `ContractSpec` class from the Stellar SDK provides a powerful way to handle and interact with the smart contract specifications. Here are some of the key functionalities:

Author: Yes it is but i wasn't sure if i should use that since i used import Spec as contractSpec, but i'll use the new method.

        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
        balance = balance.checked_add(amount).ok_or(TokenManagerError::OverflowError)?;
        e.storage().set(&key, &balance);
        Ok(())
    }

    pub fn withdraw(e: Env, user: Address, amount: i128) -> Result<(), TokenManagerError> {
        let key = Symbol::short("balance").join(&user);
        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
        if balance < amount {
            return Err(TokenManagerError::InsufficientBalanceError);
        }
        balance -= amount;
        e.storage().set(&key, &balance);
        Ok(())
    }

    pub fn get_balance(e: Env, user: Address) -> i128 {
        let key = Symbol::short("balance").join(&user);
        e.storage().get(&key).unwrap_or(0)
    }
}
```

##### Step 1: Deploy the Contract

First, deploy your Soroban smart contract to the network.

```bash
soroban contract deploy --source orbit --network testnet --wasm token_manager.wasm
```

This command will return the contract ID, for example: `CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA`.

##### Step 2: Generate TypeScript Bindings

> Use the Soroban CLI to generate the TypeScript bindings which include the contract specification.

Reviewer: Suggestion: a little more detail
```suggestion
Use the Soroban CLI to generate TypeScript bindings which use the contract's schema specifications to provide type-checked contract methods.
```

```bash
soroban contract bindings typescript --output-dir testbindings --network testnet --contract-id CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA
```

This command will create TypeScript bindings in the specified directory.

> Alternatively, you can get your spec using javascript: credit to [@mason in Stellar Developers discord](https://discord.com/channels/897514728459468821/966788672164855829/1201708320336138271) and [his gist here] (https://gist.github.com/masonforest/27ebfafa3452227937b842c56731d817)

Reviewer: nit:
```suggestion
Alternatively, you can get your spec using pure JavaScript: credit to [@mason in Stellar Developers discord](https://discord.com/channels/897514728459468821/966788672164855829/1201708320336138271) and [his gist here] (https://gist.github.com/masonforest/27ebfafa3452227937b842c56731d817)
```

Reviewer: Also this can probably be updated to use the new `rpc.Server.getContractWasm` et al. methods introduced in https://github.com/stellar/js-stellar-sdk/pull/960

Reviewer: Since clients won't always have the WASM available.

Reviewer: This could probably be a `Step 2a` or `Step 2 (Bonus)` section to separate it from the main method.

Reviewer: ^ still applies

  const wasmModule = await WebAssembly.compile(wasmBuffer);
  const reader = new XdrReader(
    Buffer.from(
      WebAssembly.Module.customSections(wasmModule, "contractspecv0")[0],
    ),
  );
  const specs = [];

  do {
    specs.push(xdr.ScSpecEntry.read(reader));
  } while (!reader.eof);

  return new ContractSpec(specs);
}

async function main() {
  const wasmFilePath = process.argv[2];

  if (!wasmFilePath) {
    console.error("Usage: node readContractSpec.js <wasmFilePath>");
    process.exit(1);
  }

  try {
    const contractSpec = await readContractSpec(wasmFilePath);
    console.log("Contract Specification:", contractSpec);
  } catch (error) {
    console.error("Error reading contract specification:", error);
  }
}

main();
```

### Running the Script

To run the script with `mycontract.wasm`, use the following command in your terminal:

```bash
node readContractSpec.js mycontract.wasm
```

This will execute the `readContractSpec` function with the `mycontract.wasm` file provided as a command-line argument and output the contract specification to the console. Copy the output for the next step

##### Step 3: Extract the Contract Spec

> Inside the generated bindings directory (`testbindings/src/index.ts`)(if you used the generated bindings), you'll find the contract specification. Look for a line that initializes a new `ContractSpec` object, such as: If you used the easy javascript then just use the output from that!

Reviewer: Reminder: I think it's `contract.Spec` now, so it'd just be `Spec`

Reviewer: Also the sentence after `such as:` should come after the code snippet

```typescript
new ContractSpec(["thespecishere"]);
```

Copy the spec from there and use it in your JavaScript class.

##### Step 4: Create the Contract Client Class

Create a new file called `TokenManagerContract.ts` and define the contract client class using the extracted contract specification.

```typescript
// TokenManagerContract.ts
import { Address, Contract } from "@stellar/stellar-sdk";
import { Spec as ContractSpec, i128 } from "@stellar/stellar-sdk/contract";

// Copy the contract spec from the generated TypeScript bindings
const contractSpec = new ContractSpec([
  "AAAAAAAAAAAAAAAKaW5pdGlhbGl6ZQAAAAAABAAAAAAAAAAFYWRtaW4AAAAAAAATAAAAAAAAAAJkZXBvc2l0AAAAAAATAAAAAAAAAAl1c2VyAAAAEwAAAAAAAAAGYW1vdW50AAAAAAATAAAAAAAAAAR3aXRoZHJhdwAAAAAABgAAAAAAAAAFcmVzdWx0AAAAAAA=",
  "AAAAAAAAAAAAAAANZ2V0X2JhbGFuY2UAAAAAAAIAAAAAAAAAAl9yZXN1bHQAAAAAEwAAAAAAAAAJdXNlcl9hZGRyAAAAAAAT",

>   "AAAAAAAAAAAAAAAAAUNvbnRyYWN0RXJyb3JNYW5hZ2VyAAAACwAAAAAAAAAAAAALAA==",

Reviewer: It'd be good to clarify what this is for folks unaccustomed to seeing raw XDR. Also this could be a source of bitrot if specs change over protocols :/

]);

> export class TokenManagerContract extends Contract {

Reviewer: wait isn't this what the `contract.Client` is?

    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals(
      "get_balance",
      {
        user,
      },
    );
    const operation = this.call("get_balance", ...invokeArgs);
    return operation.toXDR("base64");
  }

  static readonly parsers = {
    initialize: (result: string) =>
      TokenManagerContract.spec.funcResToNative("initialize", result),
    deposit: (result: string) =>
      TokenManagerContract.spec.funcResToNative("deposit", result),
    withdraw: (result: string) =>
      TokenManagerContract.spec.funcResToNative("withdraw", result),
    get_balance: (result: string) =>
      TokenManagerContract.spec.funcResToNative("get_balance", result),
  };
}

export const Errors = {
  1: { message: "Internal error" },
  2: { message: "Unauthorized error" },
  3: { message: "Insufficient balance" },
  4: { message: "Overflow error" },
};
```

##### Step 5: Invoke the Contract

To invoke a smart contract function, create an instance of the contract client and call the desired function with the appropriate arguments.

```typescript
import { Address } from "@stellar/stellar-sdk";
import { TokenManagerContract } from "./TokenManagerContract";

> import { invokeSorobanOperation, TxParams } from "./invokeSorobanOperation"; // Assuming you have these utility functions

Reviewer: A quick explanation on what these do would be great!

      minTime: 0,
      maxTime: 0,
    },
    networkPassphrase: config.passphrase,
  },
  signerFunction: async (txXdr: string) => {
    return signWithKeypair(txXdr, config.passphrase, config.admin);
  },
};

// Invoke the initialize function
const adminAddress = Address.fromString(
  "GB4UG5H5MCYPIH7GM7CBZBLOOIE2K7HMDMPK4TPZ5UEIZ2HOI5RL535V",
);

const txXDR = tokenManager.initialize(adminAddress);
console.log("Transaction XDR:", txXDR);

// Example invocation
const result = await invokeSorobanOperation(
  txXDR,
  TokenManagerContract.parsers.initialize,
  txParams,
);
console.log("Result:", result);
```

### Full TokenManagerContract.ts File

```typescript
// TokenManagerContract.ts
import { Address, Horizon, Contract } from '@stellar/stellar-sdk';
import { Spec as ContractSpec, i128 } from '@stellar/stellar-sdk/contract';

/**
 * Interface representing a network configuration.
 */
export interface Network {
  rpc: string;
  passphrase: string;
  maxConcurrentRequests?: number;
  opts?: Horizon.Server.Options;
}

/**
 * Object containing network configurations.
 */
export const networks = {
  testnet: {
    networkPassphrase: "Test SDF Network ; September 2015",

>     contractId: "CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN
> 
> 5GDDQQSX3UHYXA",

Reviewer: typo?
```suggestion
    contractId: "CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA",
```

    get_balance: (result: string) => TokenManagerContract.spec.funcResToNative('get_balance', result),
  };

  constructor(address: string) {
    super(address);
  }

  public initialize(admin: Address) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals('initialize', {
      admin,
    });
    const operation = this.call('initialize', ...invokeArgs);
    return operation.toXDR('base64');
  }

  public deposit(user: Address, amount: i128) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals('deposit', {
      user,
      amount,
    });
    const operation = this.call('deposit', ...invokeArgs);
    return operation.toXDR('base64');
  }

  public withdraw(user: Address, amount: i128) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals('withdraw', {
      user,
      amount,
    });
    const operation = this.call('withdraw', ...invokeArgs);
    return operation.toXDR('base64');
  }

  public get_balance(user: Address) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals('get_balance', {
      user,
    });
    const operation = this.call('get_balance', ...invokeArgs);
    return operation.toXDR('base64');
  }
}
```

> ### Helper Function: invokeSorobanOperation

Reviewer: oh nice nevermind, but an explanation would still be nice!

import { invokeSorobanOperation, TxParams } from "./invokeSorobanOperation"; // Assuming you have these utility functions

// Create an instance of the contract client
const contractAddress =
  "CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA";
const tokenManager = new TokenManagerContract(contractAddress);

// Define transaction parameters
const txParams: TxParams = {
  account: await config.rpc.getAccount(config.admin.publicKey()),
  txBuilderOptions: {
    fee: "10000",
    timebounds: {
      minTime: 0,
      maxTime: 0,
    },
    networkPassphrase: config.passphrase,
  },
  signerFunction: async (txXdr: string) => {
    return signWithKeypair(txXdr, config.passphrase, config.admin);
  },
};

// Invoke the initialize function
const adminAddress = Address.fromString(
  "GB4UG5H5MCYPIH7GM7CBZBLOOIE2K7HMDMPK4TPZ5UEIZ2HOI5RL535V",
);

const txXDR = tokenManager.initialize(adminAddress);
console.log("Transaction XDR:", txXDR);

// Example invocation
const result = await invokeSorobanOperation(
  txXDR,
  TokenManagerContract.parsers.initialize,
  txParams,
);
console.log("Result:", result);
```

### Conclusion

> By following these steps, you can create a contract client and interact with Soroban smart contracts using JavaScript. This guide covered the basics of obtaining the contract specification, creating a contract client, and invoking contract functions.

Reviewer: Call to action and/or cross-references on how to keep going from here would be awesome!

Reviewer: ^ still applies!

    UnauthorizedError = 2,
    InsufficientBalanceError = 3,
    OverflowError = 4,
}

#[contractimpl]
impl TokenManager {
    pub fn initialize(e: Env, admin: Address) {
        e.storage().set(&Symbol::short("admin"), &admin);
    }

    pub fn deposit(e: Env, user: Address, amount: i128) -> Result<(), TokenManagerError> {
        let key = Symbol::short("balance").join(&user);
        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
        balance = balance.checked_add(amount).ok_or(TokenManagerError::OverflowError)?;
        e.storage().set(&key, &balance);
        Ok(())
    }

    pub fn withdraw(e: Env, user: Address, amount: i128) -> Result<(), TokenManagerError> {
        let key = Symbol::short("balance").join(&user);
        let mut balance: i128 = e.storage().get(&key).unwrap_or(0);
        if balance < amount {
            return Err(TokenManagerError::InsufficientBalanceError);
        }
        balance -= amount;
        e.storage().set(&key, &balance);
        Ok(())
    }

    pub fn get_balance(e: Env, user: Address) -> i128 {
        let key = Symbol::short("balance").join(&user);
        e.storage().get(&key).unwrap_or(0)
    }
}
```

> There are a few options to get your contract Spec object to work with in javascript. The Spec is constructed from a (xdr.ScSpecEntry)[https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/xdr/curr/Stellar-contract-spec.x#L228]

Reviewer: ```suggestion
There are a few options to get your contract `Spec` object to work with in JavaScript. The `Spec` is constructed from an (xdr.ScSpecEntry)[https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/xdr/curr/Stellar-contract-spec.x#L228]
```

#### Option 1: Create Contract Specification in JavaScript

> To create a contract specification from scratch in JavaScript, we need to define the functions and errors of the contract. Here's how you can do it:

Reviewer: Should we be recommending this method? I feel like nobody should do this and should instead leverage 2(a) as the best "manual" option. Maybe this method should be last? i.e. ordered by usefulness/popularity

```typescript

> import { xdr } from "@stellar/stellar-base";
> import { Spec } from "@stellar/stellar-sdk";

Reviewer 2: `Spec` needs to be imported from the `contract` entrypoint:
```suggestion
import { xdr } from "@stellar/stellar-sdk";
import { Spec } from "@stellar/stellar-sdk/contract";
```

...or used from the `contract` import:
```suggestion
import { xdr, contract } from "@stellar/stellar-sdk";
// later, can use `contract.Spec` instead of just `Spec`
```

#### Option 2 Get the Spec from a Deployed Contract.

##### Step 1: Deploy the Contract

First, deploy your Soroban smart contract to the network.

```bash
soroban contract deploy --source orbit --network testnet --wasm token_manager.wasm
```

This command will return the contract ID, for example: `CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA`.

##### Step 2: Generate TypeScript Bindings

Use the Soroban CLI to generate TypeScript bindings which use the contract's schema specifications to provide type-checked contract methods.

```bash
soroban contract bindings typescript --output-dir testbindings --network testnet --contract-id CBEUVPRZKMCX5MPUAF2JMKMRIFEXXJXQXBLUUW6LMN5GDDQQSX3UHYXA
```

This command will create TypeScript bindings in the specified directory.

Alternatively, you can get your spec using javascript: credit to [@mason in Stellar Developers discord](https://discord.com/channels/897514728459468821/966788672164855829/1201708320336138271) and [his gist here] (https://gist.github.com/masonforest/27ebfafa3452227937b842c56731d817)

Save this file as `readContractSpec.js`

```typescript
import fs from "node:fs/promises";

> import { Contract, xdr } from "@stellar/stellar-sdk";
> import jsXdr from "@stellar/js-xdr";
> 
> const { XdrReader } = jsXdr;

Reviewer: I think you can actually pull `XdrReader` from the SDK because it's exported by base!
```suggestion
import { Contract, xdr, XdrReader } from "@stellar/stellar-sdk";
```

async function readContractSpec(path) {
  const wasmBuffer = await fs.readFile(path);
  const wasmModule = await WebAssembly.compile(wasmBuffer);
  const reader = new XdrReader(
    Buffer.from(
      WebAssembly.Module.customSections(wasmModule, "contractspecv0")[0],
    ),
  );
  const specs = [];

  do {
    specs.push(xdr.ScSpecEntry.read(reader));
  } while (!reader.eof);

  return new Contract.Spec(specs);
}

async function main() {
  const wasmFilePath = process.argv[2];

  if (!wasmFilePath) {
    console.error("Usage: node readContractSpec.js <wasmFilePath>");
    process.exit(1);
  }

  try {
    const contractSpec = await readContractSpec(wasmFilePath);
    console.log("Contract Specification:", contractSpec);
  } catch (error) {
    console.error("Error reading contract specification:", error);
  }
}

main();
```

### Running the Script

To run the script with `mycontract.wasm`, use the following command in your terminal:

```bash
node readContractSpec.js mycontract.wasm
```

> This will execute the `readContractSpec` function with the `mycontract.wasm` file provided as a command-line argument and output the contract specification to the console. Copy the output for the next step

Reviewer: Instead of making it a standalone CLI tool, you could just incorporate it into the next step as a code-only step.

#### Step 3: Extract the Contract Spec from Generated Bindings

> Inside the generated bindings directory (`testbindings/src/index.ts`)(if you used the generated bindings), you'll find the contract specification. Look for a line that initializes a new `Spec` object, or `ContractSpec`, such as: If you used the easy javascript then just use the output from that!

Reviewer: > easy javascript
?

  }

  public withdraw(user: Address, amount: i128) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals("withdraw", {
      user,
      amount,
    });
    const operation = this.call("withdraw", ...invokeArgs);
    return operation.toXDR("base64");
  }

  public get_balance(user: Address) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals(
      "get_balance",
      {
        user,
      },
    );
    const operation = this.call("get_balance", ...invokeArgs);
    return operation.toXDR("base64");
  }

  static readonly parsers = {
    initialize: (result: string) =>
      TokenManagerContract.spec.funcResToNative("initialize", result),
    deposit: (result: string) =>
      TokenManagerContract.spec.funcResToNative("deposit", result),
    withdraw: (result: string) =>
      TokenManagerContract.spec.funcResToNative("withdraw", result),
    get_balance: (result: string) =>
      TokenManagerContract.spec.funcResToNative("get_balance", result),
  };
}

export const Errors = {
  1: { message: "Internal error" },
  2: { message: "Unauthorized error" },
  3: { message: "Insufficient balance" },
  4: { message: "Overflow error" },
};

> ```

Reviewer: It'd be good to mention here that I'm pretty sure that this is what the `Client` does (cc @chadoh?) when you use the bindings!

  };

  constructor(address: string) {
    super(address);
  }

  public initialize(admin: Address) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals(
      "initialize",
      {
        admin,
      },
    );
    const operation = this.call("initialize", ...invokeArgs);
    return operation.toXDR("base64");
  }

  public deposit(user: Address, amount: i128) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals("deposit", {
      user,
      amount,
    });
    const operation = this.call("deposit", ...invokeArgs);
    return operation.toXDR("base64");
  }

  public withdraw(user: Address, amount: i128) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals("withdraw", {
      user,
      amount,
    });
    const operation = this.call("withdraw", ...invokeArgs);
    return operation.toXDR("base64");
  }

  public get_balance(user: Address) {
    const invokeArgs = TokenManagerContract.spec.funcArgsToScVals(
      "get_balance",
      {
        user,
      },
    );
    const operation = this.call("get_balance", ...invokeArgs);
    return operation.toXDR("base64");
  }

> }

Reviewer: Isn't this a repeat of the prior section?
